src/RecordComponent.cpp

```diff
@@ -215,12 +225,18 @@ void RecordComponent::flush(
         /*
          * This catches when a user forgets to use resetDataset.
          */
-        if (rc.m_dataset.dtype == Datatype::UNDEFINED)
-        {
-            throw error::WrongAPIUsage(
-                "[RecordComponent] Must set specific datatype (Use "
-                "resetDataset call).");
-        }
+        /*
+         * Suggestion for keeping this check:
+         * Use a std::optional<Dataset> in RecordComponentData, throw an error
+         * if it is still empty here, let users explicitly set Undefined
+         * datatypes for backend-specific actions.
+         */
+        // if (rc.m_dataset.dtype == Datatype::UNDEFINED)
+        // {
+        //     throw error::WrongAPIUsage(
+        //         "[RecordComponent] Must set specific datatype (Use "
+        //         "resetDataset call).");
+        // }
```

Comment:
Note that this check was inactive up to now, since RecordComponentData::RecordComponentData initialized that field with Datatype::CHAR. Using an optional would make these things more obvious and avoid such pitfalls.
To be done in a different PR though

Reply:
#1316 now uses `std::optional`